README.md

```diff
@@ -119,6 +123,55 @@ Serve files relative to `path`.
 Byte offset at which the stream starts, defaults to 0. The start is inclusive,
 meaning `start: 2` will include the 3rd byte in the stream.
 
+##### transform
+
+A function which returns a stream which can be used to transform the data.
+If this option is passed it will change the defaults of `etag` and `lastModified`
+to `false`.  If the transform stream changes the content length, it is also
+responsible for updating the `Content-Length` header or changing to
```

Comment:
This is likely to be a large foot gun. Is there some way to just make this work without adding such a potential issue? For example, for the intended use cases of this, is it more likely than not that the content length will be different from before?

Reply:
I thought hard on this.  I don't think there is a good solution which doesn't rely on the user setting it if they are modifying the content.  So in my case, I need to buffer the entire file to parse the json and then transform.

In this case, I can calculate the new length, but if you are doing something like escaping content in a CSV you might be able to do it mid stream and in that way you would not know the content length in advance.  In this case, you might consider moving to chunked encoding to avoid it.  One of the reference implementations just does that for you, but that is also very presumptuous.

TBQH, I feel that this is an advanced feature, requiring advanced understanding.  That justifies, to me, documenting it and letting users "do the right thing" for their use case.

Reply:
Maybe this feature is being implemented on the "wrong side" of this module's operations. As in, maybe the transformation should happen prior to the stream entering into this module, perhaps solving the virtual fs stuff folks want can also solve this as well, including keeping all the features of this module working an intact, like ranges, etags, etc.

I'm thinking like in a way, the transform is a kind of virtual fs in itself, where the contents, at least, are virtual.

Reply:
Hm, well if we allowed you to pass in the reference to `fs` then yeah we could do that.  That said, I am *strongly* opposed to virtual fs implementations.  There is so much complexity in that, it is just a really hard trade off compared to such a simple feature such as this.  Have you ever dug into things like `vinyl-fs`? Maybe there are better more modern ones, but either way it is a big dependency to take on.

Reply:
So to expand on what I mean by a virtual fs is _not_ those enormous things. More like providing an interface that is much simpler to do the little but that this module needs like "hey, give me a readable stream for the file of this name" or some such. In that way, the returned stream in this case could have all the transforms added at that point, which would be below where things like ranging and similar are done. If the "virtual" thing didn't understand (or able to support) ranges, it would just return the full stream, for example. I am just thinking out loud here, to be honest. I just think having the transform part between this module and the fs seems like a better fit to get all the features working, but happy to hear other solutions.

Reply:
Ahh, ok then yeah I think I can get on board with that.  So we would need `fs.stat` and `fs.createReadStream`.  I think it would be pretty simple to do that.  So in the case of doing a transform you would have to read the metadata and manage your own `stat` data based on the transform you are going to do.  So in my case, I would have to modify the file size in the `stat` call by reading the whole file and applying the transform, then throw it out.  I guess maybe I could cache it in memory, anyway it would push more of the complexity to the end user, but would allow you to be more technically correct.

Reply:
Right. And to also be clear: I don't think we even need to copy those names / those specific interfaces. We can "dumb it down" to what is actually needed for this module to do it's operations and let the "default impl" map to those `fs.*` calls, but when users want to do something different, they get an API that is easier and only what they need to work with. For example, `fs.stat` returns a lot of things we don't need -- so no need to make the user implement a `fs.stat` interface. The same for `fs.createReadStream` especially, as that take a whole slew of options that this module is just never going to need to send, and hammering it down to a simplified interface also means that it is clear to the user what, exactly, they need to implement to get it working well.

I know we need to fix the TOCTOU issue, so perhaps we can roll those things together. But even if not, we would do well to make this interface not prevent us from resolving that issue, at least :) To be specific, the underlying changes that were being held is to (1) open a fd to the fs path, (2) stat the fd and then (3) read from the fd. I suppose this simplified interface could provide the same mechanical interface or something.